### PR TITLE
chore: make the LayoutReader more poll-y

### PR DESCRIPTION
**Pull request by a10y**
I'm trying to sort out the reader, there's a funky mix right now of poll-y logic and internal state manipulation. 

* Rename some types and methods to be more in-line with the `poll` types that are more typical in Rusty interfaces
* IN PROGRESS: extract the mutable state out of the polling LayoutReaders to instead have it injected as arguments to poll methods

The proximal inspiration is the Future, Stream, etc. traits but also some inspiration from GlareDB: 

* [ExecutableOperator](https://github.com/GlareDB/glaredb/blob/main/crates/rayexec_execution/src/execution/operators/mod.rs#L268)
* [ExecutionPipeline](https://github.com/GlareDB/glaredb/blob/main/crates/rayexec_execution/src/execution/executable/pipeline.rs#L327)

### CODE REVIEW

File: vortex-file/src/read/layouts/flat.rs [hunk 1]
```diff
@@ -96,29 +101,30 @@ impl LayoutReader for FlatLayoutReader {
...
+                    .scan
+                    .expr
+                    .as_ref()
+                    .map(|e| e.evaluate(&selected))
+                    .transpose()?
+                    .unwrap_or(selected);
+                Ok(PollRead::Ready(Some(filtered)))
+            } else {
+                Ok(PollRead::Ready(None))
+            }
```
Comment (danking) [inline vortex-file/src/read/layouts/flat.rs:119]: I kinda want to ban transpose; it's so hard to grok.

Comment (a10y) [reply vortex-file/src/read/layouts/flat.rs:119]: I think all of these long operator chains are too hard to grok tbh. like once you have 6 things being chained, with closures etc. I can't just look at it and immediately figure out what's going on

File: vortex-file/src/read/layouts/inline_dtype.rs [hunk 2]
```diff
@@ -125,19 +126,19 @@ impl LayoutReader for InlineDTypeLayoutReader {
         self.child_reader()?.add_splits(row_offset, splits)
     }
 
-    fn read_selection(&self, selector: &RowMask) -> VortexResult<Option<BatchRead>> {
+    fn poll_read(&self, selector: &RowMask) -> VortexResult<PollRead<Option<ArrayData>>> {
```
Comment (danking) [inline vortex-file/src/read/layouts/inline_dtype.rs:129]: I like this a lot better, but maybe we could typedef an `ArrayDataRead` and a `MetadataRead`?

File: vortex-file/src/read/layouts/flat.rs (hunk 1 above)
Comment (AdamGS) [reply vortex-file/src/read/layouts/flat.rs:119]: _scala has entered the chat_

File: vortex-file/src/read/layouts/chunked.rs [hunk 3]
```diff
@@ -164,6 +164,15 @@ impl ChildRead {
...
+/// Reader for chunked arrays.
+///
+/// `ChunkedLayoutReader` is a nested layout that provides a reader for each chunk, as well as an
+/// optional "metadata layout".
+///
+/// ## Metadata Layout
+///
+/// The metadata layout provides a way to read a [table][ArrayData] of statistics. These statistics
+/// can be examined to make pruning decisions at scan time.
 #[allow(dead_code)]
```
Comment (robert3005) [inline vortex-file/src/read/layouts/chunked.rs:176]: I think this is no longer dead code

File: vortex-file/src/read/mod.rs [hunk 4]
```diff
@@ -59,28 +59,19 @@ pub type MessageId = Vec<LayoutPartId>;
 pub struct MessageLocator(pub MessageId, pub ByteRange);
 
 #[derive(Debug)]
-pub enum BatchRead {
+pub enum PollRead<T> {
```
Comment (robert3005) [inline vortex-file/src/read/mod.rs:62]: How about PollRead has `None` option as well? So you don't have `PollRead<Option<T>>` but maybe it' more obvious

**Pull request closed without merging**